test/test_broker.hpp

```diff
@@ -796,77 +819,156 @@ class test_broker {
     // TODO: Maybe change the name of this function.
     template <typename Endpoint>
     void close_proc(Endpoint&& ep, bool send_will) {
+        // Hold the lifetime of this endpoint until the function exits.
         auto spep = ep.shared_from_this();
-        auto& idx = cons_.get<tag_con>();
-        auto it = idx.find(spep);
-        if (it == idx.end()) return;
-
-        auto cs = ep.clean_session();
-        auto client_id = it->cid;
-
-        {   // will processing
-            auto it = will_.find(spep);
-            if (it != will_.end()) {
-                if (send_will) {
-                    do_publish(
-                        std::make_shared<std::string>(std::move(it->will.topic())),
-                        std::make_shared<std::string>(std::move(it->will.message())),
-                        it->will.qos(),
-                        it->will.retain(),
-                        it->will.props());
-                }
-                will_.erase(it);
-            }
-        }
 
-        // Remove the client from the clientid <-> connection map.
-        idx.erase(it);
+        auto & act_sess_idx = active_sessions_.get<tag_con>();
+        auto act_sess_it = act_sess_idx.find(spep);
+        // This assert triggers sometimes. Looks like close_proc getting called on the same connection?
+        // once for the disconnect, one for the close of the socket, maybe?
+        // Needs more investigation.
+        //BOOST_ASSERT(act_sess_it != act_sess_idx.end());
+        if(act_sess_it == act_sess_idx.end()) return; // TODO: This is probably an error?
 
+        std::string client_id;
+        mqtt::optional<mqtt::will> will;
+        if(ep.clean_session() && (ep.get_protocol_version() == mqtt::protocol_version::v3_1_1)) {
+            client_id = act_sess_it->client_id;
+            will = act_sess_it->will;
+            act_sess_idx.erase(act_sess_it);
+        }
+        else
+        {
+            session_state state;
+            act_sess_idx.modify(act_sess_it,
+                                [&](session_state & val)
+            {
+                state = std::move(val);
+            });
+            // TODO: Should yank out the messages from this connection object and store it in the session_state object??
+            mqtt::visit(
+                mqtt::make_lambda_visitor<void>(
+                    [&](auto& con) {
+                        con.reset(); // clear the shared pointer, so it doesn't stay alive after this funciton ends.
+                    }
+                ),
+                state.con
+            );
+            act_sess_idx.erase(act_sess_it);
+            BOOST_ASSERT(non_active_sessions_.get<tag_client_id>().find(state.client_id) == non_active_sessions_.get<tag_client_id>().end());
+            auto const& non_active_sess_it = non_active_sessions_.emplace_hint(non_active_sessions_.end(), std::move(state));
```

Comment:
Passing `end()` to `emplace_hint()` is not efficient. You need to use something like `lower_bound()` and value comparison.

See https://stackoverflow.com/a/41507893/1922763

Reply:
I noticed that it is from my code ;), but it is confusing. Please replace simple `emplace()` or lower_bound approach. *1